i3pystatus/github.py

```diff
@@ -1,83 +1,544 @@
         try:
-            if self.access_token:
-                response = requests.get('https://api.github.com/notifications?access_token=' + self.access_token)
-            else:
-                response = requests.get('https://api.github.com/notifications', auth=(self.username, self.password))
-            data = json.loads(response.text)
-        except (ConnectionError, Timeout) as e:
-            logger.warn(e)
-            data = []
-
-        # Bad credentials
-        if isinstance(data, dict):
-            err_msg = data['message']
-            raise ConfigError(err_msg)
-
-        format_values = dict(unread_count='', unread='')
-        unread = len(data)
-        if unread > 0:
-            format_values['unread_count'] = unread
-            format_values['unread'] = self.unread_marker
-
-        self.data = format_values
-        self.output = {
-            'full_text': self.format.format(**format_values),
-            'color': self.color
-        }
+            self.config_error = None
+            self.update_unread()
+        except ConfigError as exc:
+            self.config_error = exc
+
+        self.data['update_error'] = self.update_error \
+            if self.failed_update \
+            else ''
+        self.refresh_display()
+        if self.failed_update:
+            self.notify('Error occurred updating module. See log for details.')
+
+    @require(internet)
+    def update_status(self):
+        try:
+            # Get most recent update
+            if not hasattr(self, 'last_message_url'):
+                self.last_message_url = \
+                    self.api_request(API_METHODS_URL)['last_message_url']
+            self.current_status = self.api_request(self.last_message_url)
+            if not self.current_status:
+                self.failed_update = True
+                return
+
+            self.data['status'] = self.status.get(
+                self.current_status.get('status'),
+                self.unknown_status)
+
+            if self.previous_status is not None:
+                if self.current_status != self.previous_status:
+                    self.show_status_notification()
+            self.previous_status = self.current_status
+
+        except Exception:
+            # Don't let an uncaught exception kill the update thread
+            self.logger.error(
+                'Uncaught error occurred while checking GitHub status. '
+                'Exception follows:', exc_info=True
+            )
+            self.failed_update = True
+
+    @require(internet)
+    def update_unread(self):
+        # Reset the new_unread attribute to prevent spurious notifications
+        self.new_unread = None
+
+        try:
+            if not self.username and not self.password and not self.access_token:
+                # Auth not configured
+                self.logger.debug(
+                    'No auth configured, notifications will not be checked')
+                return False
```

Comment:
This is meant to allow potential future development choices to easily know whether or not the function was successful, e.g.

```python
if not self.update_unread():
    do_something_here
```